File: LeetCode/Medium/0039-combination-sum/0039-combination-sum.java
class Solution {
    List<List<Integer>> ret;
    int target;
    int[] candidates;
    public List<List<Integer>> combinationSum(int[] candidates, int target) {
        ret = new LinkedList();
        this.target = target;
        this.candidates = candidates;
        DFS(new ArrayList(), 0, 0);
        return ret;
    }
    
    private void DFS(ArrayList<Integer> curr, int start, int sum){
        if (sum == target) {
            ret.add(new ArrayList(curr));
            return;
        }
        if (start >= candidates.length || sum > target) return;
        
        curr.add(candidates[start]);
        DFS(curr, start, sum + candidates[start]);
        curr.remove(curr.size() - 1);
        DFS(curr, start + 1, sum);
    }
}

File: LeetCode/Medium/0039-combination-sum/0039-combination-sum.py
class Solution:
    def combinationSum(self, candidates: List[int], target: int) -> List[List[int]]:
        ret = []

        def DFS(curr, start, total):
            if total == target:
                ret.append(curr)
                return
            if start >= len(candidates) or total > target:
                return
            
            DFS(curr + [candidates[start]], start, total + candidates[start])
            DFS(curr, start+1, total)

        DFS([], 0, 0)

        return ret